Adds Usage breakdown card

a2batic opened the pull request:
![Screenshot from 2019-10-22 02-52-30](https://user-images.githubusercontent.com/12200504/67245878-b0beea00-f47a-11e9-8008-cf1602f776a2.png)
![Screenshot from 2019-10-22 02-53-58](https://user-images.githubusercontent.com/12200504/67245888-b7e5f800-f47a-11e9-9f53-e11cc2904ecc.png)
![Screenshot from 2019-10-22 02-55-04](https://user-images.githubusercontent.com/12200504/67245891-bae0e880-f47a-11e9-8b41-83e6bd42cea0.png)
![Screenshot from 2019-10-22 02-56-17](https://user-images.githubusercontent.com/12200504/67245911-c46a5080-f47a-11e9-811e-13edc53b96b5.png)

Removes top consumer card and adds capacity breakdown card
Signed-off-by: Afreen Rahman <afrahman@redhat.com>
Signed-off-by: Kanika <kmurarka@redhat.com>

Thanks @afreen23 @anmolsachan for the pointers

a2batic commented:
/assign @afreen23 @gnehapk @cloudbehl

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.scss:
```diff
@@ -0,0 +1,26 @@
+
+.ceph-capacity-breakdown-monitoring-header-link{
+    padding: 0.3rem;
+    margin-right: 1rem;
```
@a2batic two different units have been used `rem` and `px`. Any reasons for that?

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+import { PROJECTS, STORAGE_CLASSES } from '../../../../constants/index';
+import { CAPACITY_BREAKDOWN_QUERIES, CAPACITY_USAGE_QUERIES2 } from '../../../../constants/queries';
+import { Grid, GridItem } from '@patternfly/react-core';
+import { withDashboardResources, DashboardItemProps } from '@console/internal/components/dashboard/with-dashboard-resources';
```
alpha imports

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+import { getGraphLegendsStats, getGraphVectorStats, HeaderPrometheusLink } from './utils';
+import { ChartLegend } from '@patternfly/react-charts';
+import './capacity-breakdown-card.scss';
+import { TotalCapacityBody, CapacityStackBody, CapacityLegendsBody, AvailableCapacityBody } from '@console/shared/src/components/dashboard/breakdown-card/breakdown-body';
```
alpha imports

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+import { TotalCapacityBody, CapacityStackBody, CapacityLegendsBody, AvailableCapacityBody } from '@console/shared/src/components/dashboard/breakdown-card/breakdown-body';
+import { RootState } from '@console/internal/redux';
+import { getPrometheusExpressionBrowserURL } from '@console/internal/components/graphs/prometheus-graph';
+import { MonitoringRoutes, connectToURLs } from '@console/internal/reducers/monitoring';
```
alpha imports

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+const metricTypes = _.keys(capacityResourceValue);
+const metricTypesOptions = _.zipObject(metricTypes, metricTypes);
+
+
```
Please remove one extra line here.

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+      (q) => prometheusResults.getIn([CAPACITY_USAGE_QUERIES2[q], 'data']) as PrometheusResponse,
+    );
+
+
```
Please remove one extra line here.

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+      loader: () => import('./capacity-breakdown-items').then((m) => m.ResultAvailable),
+    });
+
+    const queriesLoaded2 = Object.keys(CAPACITY_BREAKDOWN_QUERIES).every(
```
can be named better

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+
+    const topConsumers: TopConsumersQueries = { ...CAPACITY_BREAKDOWN_QUERIES };
+
+    const results3 = prometheusResults.getIn([
```
same as above comment

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+      'data',
+    ]) as PrometheusResponse;
+
+    const results4 = prometheusResults.getIn([
```
same as above comment

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+    ]) as PrometheusResponse;
+
+
+
```
Remove extra lines here

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+    ));
+
+    return (
+
```
this line can be removed

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+          />
+        </GridItem>
+        {chartLegendList.map((legends) => {
+          let prometheusData2 = [];
```
can be named better

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,171 @@
+    const LegendResult = _.get(results3, 'data.result', []);
+    const chartLegendsData = getGraphVectorStats(LegendResult);
+    const legendData = getGraphLegendsStats(chartLegendsData);
+    const chartLegendList = legendData.map((data, i) => (
```
Add types for all these

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-items.tsx:
```diff
@@ -0,0 +1,82 @@
+export const ResultAvailable: React.FC<ResultAvailableProps> = ({ results }) => {
+
+    const { available } = getCardAvailableValue(results);
+
```
extra lines can be removed here

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-items.tsx:
```diff
@@ -0,0 +1,82 @@
+    const allConsumerResult = _.get(results[1], 'data.result', []);
+    const chartData = getGraphVectorStats(topConsumerStatsResult, allConsumerResult);
+
+    const chartBarList = chartData.map((data, i) => (
```
Add type

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-items.tsx:
```diff
@@ -0,0 +1,82 @@
+        <ChartBar key={i} data={data as []} barRatio={2} barWidth={45} /> // eslint-disable-line react/no-array-index-key
+    ));
+
+
```
Remove extra line

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-items.tsx:
```diff
@@ -0,0 +1,82 @@
+    return (
+        <div>
+            <div className="stack-chart-bottom" >
+                {/* <p> {chartBarList}</p> */}
```
Remove this if not needed

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+import * as _ from 'lodash';
+import { getInstantVectorStats, GetStats } from '@console/internal/components/graphs/utils';
+import { humanizeBinaryBytesWithoutB, ExternalLink } from "@console/internal/components/utils";
```
alpha imports

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+import { humanizeBinaryBytesWithoutB, ExternalLink } from "@console/internal/components/utils";
+import { RootState } from '@console/internal/redux';
+import { getPrometheusExpressionBrowserURL } from '@console/internal/components/graphs/prometheus-graph';
+import { MonitoringRoutes, connectToURLs } from '@console/internal/reducers/monitoring';
```
alpha imports

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+import './capacity-breakdown-card.scss';
+
+let capValues;
+let top5capcitySum = 0;
```
`topFiveCapacitySum` is preferrable

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+
+let capValues;
+let top5capcitySum = 0;
+export const getCardUsedValue = (results) => {
```
add type

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+let capValues;
+let top5capcitySum = 0;
+export const getCardUsedValue = (results) => {
+    const getLastStats = (response, getStats: GetStats): React.ReactText => {
```
add type for response as well

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+    return capValues;
+}
+
+export const getCardAvailableValue = (results) => {
```
add type

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+}
+
+export const getGraphVectorStats: GetStats = (response, allresponse) => {
+    let temp, temp2 = [];
```
temp and temp2 can be named better

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+    return temp2;
+};
+
+const legendName = (obj) => {
```
add type

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+    return legendValue;
+}
+
+export const getGraphLegendsStats = (response) => {
```
add type

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/utils.tsx:
```diff
@@ -0,0 +1,108 @@
+    let temp, temp2 = [];
+    response.map((r) => {
+        temp = {
+            name: legendName(r[0]),//(r[0].name ? r[0].name : r[0].storageclass), r[0].y
```
remove commented line if not needed

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/constants/queries.ts:
```diff
@@ -55,17 +57,20 @@ export const CAPACITY_USAGE_QUERIES = {
+  [StorageDashboardQuery.CEPH_CAPACITY_USED]: 'ceph_cluster_total_used_bytes',
 };
+
+export const CAPACITY_USAGE_QUERIES2 = {
```
can `CAPACITY_USAGE_QUERIES2` can be named something else i.e. without using 2

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/extensions/dashboards.ts:
```diff
@@ -1,28 +1,6 @@
 import { Extension, DashboardsExtensionProperties } from '@console/plugin-sdk';
```
Alpha imports

gnehapk commented on frontend/packages/console-shared/src/components/dashboard/breakdown-card/breakdown-body.tsx:
```diff
@@ -0,0 +1,172 @@
+        body = objects.length ? objects : <div className="text-secondary">Not available</div>;
+    }
+    return (
+        <>
```
Do we really need fragment here?

gnehapk replied:
ok. Got it. For arrays of `TotalCapacity`

afreen23 commented:
@a2batic  I think we can reduce the complexity  by not following the pattern of activity card in general components but for capacity card or top consumers-card(in 4.2) as previously mentioned.

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.scss:
```diff
@@ -0,0 +1,26 @@
+}
+
+.co-capacity-breakdown-card__body {
+    padding-top: 24px !important;
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/plugin.ts:
```diff
@@ -111,13 +111,26 @@ const plugin: Plugin<ConsumedExtensions> = [
+      position: GridPosition.MAIN,
+      loader: () =>
+        import(
+          './components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card' /* webpackChunkName: "ceph-storage-top-consumers-card" */
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.scss:
```diff
@@ -0,0 +1,26 @@
+.stack-chart-bottom {
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.scss:
```diff
@@ -0,0 +1,26 @@
+    padding-top: 24px !important;
+}
+
+.co-capacity-breakdown-card__header{
```
ack

afreen23 commented on frontend/packages/ceph-storage-plugin/src/plugin.ts:
```diff
@@ -94,6 +94,20 @@ const plugin: Plugin<ConsumedExtensions> = [
   // Ceph Storage Dashboard Main Cards
+
+  // @TODO: change the order after development
+  {
```
Breakdown card needs to be below status card.

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-capacity.tsx:
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
+import './breakdown-card.scss';
+
+export const TotalCapacityBody: React.FC<TotalCapacityBodyProps> = ({ value, classname }) => {
```
Why are we adding prop for class.
I think we will reuse the same one in noobaa with no changes.

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-body.tsx:
```diff
@@ -0,0 +1,74 @@
+import * as React from 'react';
+import { Grid, GridItem, EmptyState, Title, EmptyStateVariant } from '@patternfly/react-core';
```
Alpha imports

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-body.tsx:
```diff
@@ -0,0 +1,74 @@
+  cephTotal,
+  metricModel,
+  formatValue,
+  isLoading,
```
Alpha imports

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-chart.tsx:
```diff
@@ -0,0 +1,96 @@
+import { DataPoint } from '@console/internal/components/graphs';
+import { K8sKind } from '@console/internal/module/k8s';
+import { resourcePathFromModel } from '@console/internal/components/utils';
+import { isAvailableBar, getBarRadius } from './utils';
```
Alpha imports

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,95 @@
+  return stats.reduce((total, dataPoint) => total + dataPoint.y, 0);
+};
+
+const addOthers = (stats: StackDataPoint[], totalUsed, formatValue) => {
```
Add types for other parameters as well.

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,95 @@
+  return othersData;
+};
+
+export const addAvailable = (stats: StackDataPoint[], total, used, totalUsed, formatValue) => {
```
Types

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,95 @@
+  const availableData = {
+    x: '',
+    y: availableInBytes,
+    label: `${formatValue(availableInBytes).string}`,
```
I don't think that you need to use `string` here as template string will convert it to string only.

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,95 @@
+
+export const isAvailableBar = (index: number, length: number) => {
+  if (index === length - 1) {
+    return { fill: '#b8bbbe' };
```
This should be declared in some variable instead of hard coding the color value here

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,95 @@
+      y,
+      name: _.truncate(name, { length: 12 }),
+      link: name,
+      color: 'blue',
```
same as above comment

a2batic replied:
ack

bipuladh commented on frontend/packages/ceph-storage-plugin/src/constants/queries.ts:
```diff
@@ -49,17 +52,67 @@ export const CAPACITY_USAGE_QUERIES = {
+    model: PodModel,
+    metric: 'pod',
+    queries: {
+      [StorageDashboardQuery.PODS_BY_USED]: `(sort_desc(topk(5, (${
```
```suggestion
      [StorageDashboardQuery.PODS_BY_USED]: `(sort_desc(topk(25, (${
```
Cross verify this data with the data from K8s. To verify that these Pods actually exists. 
https://github.com/openshift/console/blob/master/frontend/public/components/dashboard/dashboards-page/overview-dashboard/top-consumers-card.tsx#L138

bipuladh replied:
@rawagner I think it's necessary here too. What do you think?

rawagner replied:
Yes, i agree. This is the same case as with top consumers in utilization card. The resource can be in prometheus still but its already gone in openshift (project/pod/... was deleted) and you dont want to link to 404

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,102 @@
+  if (index === length - 1) {
+    return { top: 3 };
+  }
+  return {};
```
```
let barRadius = {}
 if (index === 0) {
    barRadius = { bottom: 3 };
  }
  if (index === length - 1) {
    barRadius = { top: 3 };
  }
  return barRadius;
```

a2batic replied:
ack

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,102 @@
+  return {};
+};
+
+export const isAvailableBar = (index: number, length: number) => {
```
same as above.

a2batic replied:
ack

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-card.scss:
```diff
@@ -0,0 +1,53 @@
+}
+
+.skeleton-stack-no-show {
+  background-color: $pf-color-white;
```
```suggestion
  background-color: var(--pf-global--BackgroundColor--100)
```
Remove the import

a2batic replied:
ack

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,102 @@
+  const availableData = {
+    x: '',
+    y: availableInBytes,
+    label: `${formatValue(availableInBytes).string}`,
```
```suggestion
    label: formatValue(availableInBytes).string,
```
@a2batic I think this is what @gnehapk meant here.

a2batic replied:
ack

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.tsx:
```diff
@@ -0,0 +1,78 @@
+    return () => queryKeys.forEach((key) => stopWatchPrometheusQuery(queries[key]));
+  }, [watchPrometheus, stopWatchPrometheusQuery, metricType, queryKeys, queries]);
+
+  const results = queryKeys.map((key) => prometheusResults.getIn([queries[key], 'data']));
```
It would be better if you divide the results into multiple variables so that it's easy to understand what the different queries are returning.

a2batic replied:
Is it fine if we can take it up in a follow up PR ?

bipuladh replied:
sure

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,102 @@
+  return stats.reduce((total, dataPoint) => total + dataPoint.y, 0);
+};
+
+const addOthers = (stats: StackDataPoint[], totalUsed: string, formatValue: Humanize) => {
```
```suggestion
const addOthers = (stats: StackDataPoint[], totalUsed: string, humanize: Humanize) => {
```
I think this is the convention through the console. For other instances also.

a2batic replied:
ack

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/capacity-breakdown/capacity-breakdown-card.scss:
```diff
@@ -0,0 +1,7 @@
+.ceph-capacity-breakdown-card__body {
+    padding-top: 24px;
```
Have you tried with spacers already provided by Patternfly?

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-header.tsx:
```diff
@@ -0,0 +1,14 @@
+  </div>
+);
+
+export const HeaderPrometheusLink = connectToURLs(MonitoringRoutes.Prometheus)(
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-capacity.tsx:
```diff
@@ -0,0 +1,11 @@
+type TotalCapacityBodyProps = {
+  value: string;
+  classname?: string;
+};
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-capacity.tsx:
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
+import './breakdown-card.scss';
+
+export const TotalCapacityBody: React.FC<TotalCapacityBodyProps> = ({ value, classname }) => {
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-capacity.tsx:
```diff
@@ -0,0 +1,11 @@
+import './breakdown-card.scss';
+
+export const TotalCapacityBody: React.FC<TotalCapacityBodyProps> = ({ value, classname }) => {
+  return <p className={`capacity-breakdown-card__capacity-body ${classname || ''}`}>{value}</p>;
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-card.scss:
```diff
@@ -0,0 +1,53 @@
+}
+
+.capacity-breakdown-card__capacity-body {
+  font-size: 0.9rem;
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-card.scss:
```diff
@@ -0,0 +1,53 @@
+}
+
+.capacity-breakdown-card__header-link {
+  padding: 0.3rem;
```
ack

a2batic commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-card.scss:
```diff
@@ -0,0 +1,53 @@
+
+.capacity-breakdown-card__header-link {
+  padding: 0.3rem;
+  margin-right: 1rem;
```
ack

a2batic commented:
@bipuladh Thanks for the review. Addressed all comments expect:-
1. https://github.com/openshift/console/pull/2920#discussion_r338634527
2. https://github.com/openshift/console/pull/2920#discussion_r338644773
Please review. If it's fine we can take the above in a followup PR.

bipuladh commented:
/lgtm

cloudbehl commented:
@rawagner can you review this PR.

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-capacity.tsx:
```diff
@@ -0,0 +1,11 @@
+import './breakdown-card.scss';
+
+export const TotalCapacityBody: React.FC<TotalCapacityBodyProps> = ({ value, classname = '' }) => {
+  return <p className={`capacity-breakdown-card__capacity-body ${classname}`}>{value}</p>;
```
use `classNames`
```suggestion
  return <p className={classNames('capacity-breakdown-card__capacity-body', className')}>{value}</p>;
```

and you also dont have to define default value for classname.

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-capacity.tsx:
```diff
@@ -0,0 +1,11 @@
+
+type TotalCapacityBodyProps = {
+  value: string;
+  classname?: string;
```
```suggestion
  className?: string;
```

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-card.scss:
```diff
@@ -0,0 +1,52 @@
+$skeleton-animation: loading-skeleton 1s ease-out .15s infinite alternate;
```
you should import these from `console-shared/styles/skeleton-screen`

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-chart.tsx:
```diff
@@ -0,0 +1,96 @@
+import { isAvailableBar, getBarRadius } from './utils';
+import './breakdown-card.scss';
+
+const LinkableLegend = (props) => {
```
missing type

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-chart.tsx:
```diff
@@ -0,0 +1,96 @@
+import { isAvailableBar, getBarRadius } from './utils';
+import './breakdown-card.scss';
+
+const LinkableLegend = (props) => {
```
```suggestion
const LinkableLegend = ({ metricModel, datum }) => {
```

and then the return can be inlined

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-chart.tsx:
```diff
@@ -0,0 +1,96 @@
+const LinkableLegend = (props) => {
+  const { metricModel, datum } = props;
+  return (
+    <>
```
no need for `<>`

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-chart.tsx:
```diff
@@ -0,0 +1,96 @@
+
+export const BreakdownChart: React.FC<BreakdownChartProps> = ({ data, legends, metricModel }) => {
+  const chartData = data.map((d: DataPoint, index) => (
+    <ChartBar
```
should have a `key` prop

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-header.tsx:
```diff
@@ -0,0 +1,12 @@
+import { ExternalLink } from '@console/internal/components/utils';
+import { getPrometheusExpressionBrowserURL } from '@console/internal/components/graphs/prometheus-graph';
+
+const HeaderPrometheusViewLink = ({ link, urls }) => (
```
missing type

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-loading.tsx:
```diff
@@ -0,0 +1,22 @@
+import { Grid, GridItem } from '@patternfly/react-core';
+import './breakdown-card.scss';
+
+export const BreakdownChartLoading: React.FC<{}> = () => {
```
```suggestion
export const BreakdownChartLoading: React.FC = () => {
```

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-loading.tsx:
```diff
@@ -0,0 +1,22 @@
+import './breakdown-card.scss';
+
+export const BreakdownChartLoading: React.FC<{}> = () => {
+  return (
```
no need for explicit return

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,104 @@
+import { PrometheusResponse } from '@console/internal/components/graphs';
+import { Colors } from './bar-colors';
+
+const getTotal = (stats: StackDataPoint[]) => {
```
can be one-liner

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,104 @@
+};
+
+export const getStackChartStats: GetStackStats = (response, metric, humanize) => {
+  const results = _.get(response, 'data.result', []);
```
it would be better to reuse `getInstantVectorStats` to keep the parsing logic in one place and then do another `map` to alter/add your own props IMO

a2batic replied:
ack

rawagner commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,104 @@
+  color: string;
+};
+
+export const getCapacityValue = (cephUsed: any, cephTotal: any, humanize: Humanize) => {
```
`cephUsed` and `cephTotal` should have better types than `any`.
missing return type

a2batic replied:
ack

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/breakdown-chart.tsx:
```diff
@@ -0,0 +1,94 @@
+import './breakdown-card.scss';
+
+const LinkableLegend = (props: any) => {
+  return (
```
`return` statement can be skipped here

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/breakdown-card/utils.tsx:
```diff
@@ -0,0 +1,101 @@
+};
+
+export const getStackChartStats: GetStackStats = (response, humanize) => {
+  return response.map((r, i) => {
```
`return` can be skipped here

a2batic replied:
ack

cloudbehl commented:
/lgtm
/approve

rawagner commented:
/lgtm
